### PR TITLE
Features/benchmarking task

### PR DESCRIPTION
**Pull request by steffencruz**
- Adds a benchmarking task (multiple choice). 
- Adds unit tests (900+)
- Minor refactoring of some data structures such as `REWARD_MODELS` registry.
- Adds new exception `TaskCreationError` which is raised if the generated text for a task does not meet acceptance criteria. I see this as a way of doing inline quality control (e.g. catching hallucinations based on heuristics).

I opted for few shot (system) prompting to get the the model to follow the required multiple choice formatting requirements. The reference answer is created during the query as one of the options and is marked with a *. Some examples (from the system prompt) are shown below. 

### Example 1

What is the capital of Texas?

A. Paris
B. London
*C. Austin
D. Houston

### Example 2

Which of the following best describes the primary driving force behind protein folding?

A. Covalent bond formation between amino acids
*B. Hydrophobic interactions between nonpolar side chains
C. Hydrogen bonds between the protein backbone and side chains
D. Ionic interactions between charged side chains

## Further information

Seems reasonably stable, and the subnet 1 miner scores indicate that it's decently difficult (bear in mind that they are VERY good at wiki RAG and so this is an open book exam for them). There's also a (simple) path to modulating the difficulty of this by adding a modifier to the generation prompt such as "ask an expert/basic question...".

I do not claim that this is a stable branch. It's just a demo, for now. We need to generate much more data and run a full eval pipeline on results to be sure it's high quality.

**Example data**: https://smooth-lamprey-680.notion.site/SAMPLE-f06c5b67f2f24e8c9d8514824f692960 
**Tracked experiment**: https://wandb.ai/opentensor-dev/alpha-validators/runs/eeueemz8/overview

### CODE REVIEW

File: prompting/tasks/benchmark.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+Which of the following best describes the primary driving force behind protein folding?
+
+A. Covalent bond formation between amino acids
+*B. Hydrophobic interactions between nonpolar side chains
+C. Hydrogen bonds between the protein backbone and side chains
+D. Ionic interactions between charged side chains
+"""
+
+# Used to obtain the query (which is a question about the context)
+# TODO: modulate difficulty "ask an {expert} question"
```
Comment (RusticLuftig) [inline prompting/tasks/benchmark.py:42]: This would be great! Perhaps something like "ask a question that has a X/5 difficulty" so we can get even more variety?

File: prompting/tasks/benchmark.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+import bittensor as bt
+from dataclasses import dataclass
+from prompting.tasks import Task
+from prompting.utils.exceptions import TaskCreationError
+
+# TODO: introduce criteria for the query and reference answer (length, layout, etc.) and make these arguments
+
+# Used to instruct the LLM to provide a good query when given a context
+QUERY_SYSTEM_PROMPT = """\
+You are a multiple choice question-generating expert, focusing on delivering comprehensive and accurate questions with depth and clarity for AI benchmarking purposes. The questions you generate should be specific and based on the context that is provided. Do not add any greetings or assistant messages before or after the question (e.g. Here's a question based on the context) as this does not bring any benefit to the question. The question should be self-contained. You will adhere to a word limit of 100 words for each question.
```
[the remark this answers is not in the document]
Comment (RusticLuftig) [reply prompting/tasks/benchmark.py:11]: Nevermind. I confused string length with word count. Ignore this.

File: prompting/rewards/multiple_choice.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+
+class MultipleChoiceModel(BaseRewardModel):
+    @property
+    def name(self) -> str:
+        return "multiple_choice"
+
+    def __init__(self, **kwargs):
+        super().__init__()
+        self.choices = ("A", "B", "C", "D")
```
[the remark this answers is not in the document]
Comment (steffencruz) [reply prompting/rewards/multiple_choice.py:16]: Slightly, but my gut tells me it's more likely out of distribution for the type of task which may bring more pain..

Also, we are taking the **last** occurrence of `{'A','B','C','D'}` which should make the incidental 'A' match condition less likely. 

Maybe a specific delimiter is a better way such as `A)`. There's a tradeoff between robustness and flexibility -- we don't want to throw away good answers because they don't conform

File: prompting/tasks/benchmark.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+In addition to the question, you will provide 4 possible answers (multiple choice-style), one of which is correct. Ensure that the correct answer is not too obvious, and that the incorrect answers are plausible. Be careful to maintain a neutral tone in the candidate answers, and randomly order the answers so that the correct answer is not always in the same position. Do not add any additional information after the answers.
```
[the remark this answers is not in the document]
Comment (steffencruz) [reply prompting/tasks/benchmark.py:13]: Good call. From the tracked experiment, 4/209 of the queries had a repeated option.

```
In [132]: df['query'].apply(lambda text: len(set(re.findall(r"\n\*?[A-D]\.\s+(.*)", text)))).value_counts()
Out[132]: 
query
4    205
3      4
Name: count, dtype: int64
```

[the remark this answers is not in the document]
Comment (steffencruz) [reply prompting/tasks/benchmark.py:13]: We can add both an extra instruction and a further acceptance criteria which won't allow these ones to be used.

File: prompting/rewards/multiple_choice.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+            # remove any punctuation too.
+            matches = [
+                word
+                for word in re.sub(r"\W", " ", completion).split()
+                if word in classes
+            ]
+
+            # Take the last match as the answer
+            if matches:
+                reward = matches[-1] == reference
```
Comment (Hollyqui) [inline prompting/rewards/multiple_choice.py:39]: why do we even allow the miners to return multiple responses in the first place? Should we not try and incentivise only one clear answer scoring 0 if len(matches) == 0?

```
reward = 0 if len(matches) > 1 else matches[0] == reference
```

File: prompting/rewards/pipeline.py [hunk 6]
```diff
@@ -1,27 +1,7 @@
...
-    "rouge": RougeRewardModel,
-    "relevance": RelevanceRewardModel,
-    "diff": DiffRewardModel,
-    "float_diff": FloatDiffModel,
-    "date": DateRewardModel,
-    "ordinal": OrdinalRewardModel,
-    "streaming": StreamingRewardModel,
-}
-
+from ..rewards import BaseRewardModel, REWARD_MODELS
```
Comment (Hollyqui) [inline prompting/rewards/pipeline.py:4]: Generally relative imports are discouraged in python - it's not as strongly discouraged as it once was but good to keep in mind. This happens a lot throughout the codebase so I won't comment on it every time, but it's good to keep in mind for future development!

File: prompting/tasks/benchmark.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Hollyqui) [reply prompting/tasks/benchmark.py:15]: For such low rates I would recommend checking whether a duplicate answer exists and simply regenerating rather than changing the prompt. Overall, that ends up somewhat cheaper and gives you 100% confidence that now answer is ever repeated. You can of course still modify the prompt slightly for extra safety, but with LLMs the more hard checks you can do the better, prompt engineering is always somewhat of a mess.

File: prompting/tasks/benchmark.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+        dict(name="multiple_choice", weight=1.0),
+    ]
+    penalty_definition = []
+    cleaning_pipeline = []
+
+    static_reference = True
+    challenge_type = "query"
+
+    # specific pattern (semi-flexible) which detects multiple choices
+    choices_pattern = r"\n\s*(\*?\s*\W?[A-D]\W?)\s*(.*)"
```
Comment (Hollyqui) [inline prompting/tasks/benchmark.py:69]: My general recommendation here would be to add an example to the prompt - having a flexible regex is still useful, but typically a concrete example improves quality and consistency

File: prompting/utils/exceptions.py [hunk 8]
```diff
@@ -4,3 +4,11 @@ class MaxRetryError(Exception):
     def __init__(self, message="Maximum number of retries exceeded"):
         self.message = message
         super().__init__(self.message)
+
+
+class TaskCreationError(Exception):
+    """Exception raised when the task creation fails."""
```
Comment (Hollyqui) [inline prompting/utils/exceptions.py:10]: Abstraction seems a little bit superfluous - it doesn't hurt anyone to keep it but all it does is set a default message which is only ever used in the tests

**Review by Hollyqui — APPROVED** [review]
LGTM, check comments for feedback/suggestions

File: prompting/rewards/multiple_choice.py (hunk 5 above)
Comment (steffencruz) [reply prompting/rewards/multiple_choice.py:39]: The main motivation is to prevent accidental matches.  Consider the following
```
What is the name of a baby dog?

A. kitten
B. kid
*C. puppy
D. dinosaur
```
Response:
`"A baby dog is called a puppy. The answer is A."`

This should be accepted.

Comment (steffencruz) [reply prompting/rewards/multiple_choice.py:39]: if we are case insensitive, there are actually 3 matches in the response: `A, a, C`

File: prompting/tasks/benchmark.py (hunk 7 above)
Comment (steffencruz) [reply prompting/tasks/benchmark.py:69]: I believe the systm prompt has a concrete example with the required layout.

File: prompting/rewards/multiple_choice.py (hunk 5 above)
Comment (Hollyqui) [reply prompting/rewards/multiple_choice.py:39]: Yup, but this then seems like an issue with the regex being insufficient. What if:

"Answer C is correct. A baby dog is called a puppy"

The order is arbitrary as far as I understand

File: prompting/rewards/pipeline.py (hunk 6 above)
Comment (Hollyqui) [reply prompting/rewards/pipeline.py:4]: Fair enough - my condolences. I've been there

**Comment by steffencruz** [comment]
## GPT-4 Eval Results
The quality of the benchmarking task was evaluated using GPT-4 (tracked experiment ...). I performed 5 evals on each sample, contradiction rates were small/negligible (see last plot).

The two types of evals are:
1. Whether the multiple choice question is 'good' (BUILTIN::1)
2. Whether the reference answer is correct (BUILTIN::4)

### Conclusions 
~90% of the benchmarking questions were good (scored 3+ out of max score 4) and around the same had reference answers that GPT agreed with.


### Methodology
I used the following eval prompt:
```
DEFAULT_EVAL_SYSTEM_PROMPT =  """
You are an AI assistant which judges the quality of other LLM generations.

The user will provide you with a prompt and an LLM completion. Your task is to evaluate the quality of the completion given the user prompt. You can use the following scale: 1 - bad, 2 - ok, 3 - good, 4 - excellent. An excellent completion is one which follows the instruction perfectly. This means: it contains only the requested information and nothing more, it is coherent, convincing and well-written without any unrequested placeholders, tags or otherwise discursive and off topic language or content. Any completion containing artefacts such as system messages (e.g. "Sure, I can do that for you") should be penalized appropriately.

Do not provide any information other than the quality of the completion, expressed as a number between 1 and 4.
"""
```

Then I prompt gpt-4 to score each of the benchmarking tasks. it's not perfect -- I don't check whether the questions are of educational value for example, and i don't show the block of text that was the basis for the question (the context). These would improve the results, but in terms of the overall instruction following of the LLM (i.e. has it made a decent benchmarking task) the eval results suggest that the data is quite good.


### Results
image

image

### Contradiction Rates
newplot - 2024-07-25T191259 736

File: prompting/tasks/benchmark.py [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+        """
+        # shuffle the choices
+        shuffled_choice_list = list(
+            zip("ABCD", np.random.choice(choices, len(choices), replace=False))
+        )
+
+        # match the reference and get the letter
+        new_reference = [c[0] for c in shuffled_choice_list if c[1] == reference_label][
+            0
+        ]
```
Comment (dbobrenko) [inline prompting/tasks/benchmark.py:145]: It loops through the list of all choices, even if reference was already found, minor optimization:
```suggestion
        new_reference = next(label for label, choice in shuffled_choice_list if choice == reference_label)
```

File: prompting/rewards/multiple_choice.py (hunk 5 above)
Comment (dbobrenko) [reply prompting/rewards/multiple_choice.py:39]: What if we force to accept answer in some specific tags, eg `<ANSWER>C</ANSWER>`? Not ideal solution, but it will minimize the probability of accidental matches at least

**Review by dbobrenko — APPROVED** [review]
LGTM, nicely done! Left some minor comments that can be ignored for now

**Comment by steffencruz** [comment]
Yeah @dbobrenko i agree that the way we parse the miner responses is currently the most brittle part of the design. I’m happy to choose a different approach for parsing. We could even return raw logits
